ziti/src/main/kotlin/org/openziti/identity/Enroller.kt

```diff
@@ -56,162 +87,121 @@ class Enroller(
-            throw IllegalArgumentException(error.get("message")?.toString() ?: "unknown error" )
-        } else {
-            val ct = conn.getHeaderField("Content-Type").lowercase()
-            val certs = when (ct) {
-                "application/x-pem-file" -> {
-                    readCerts(conn.inputStream.reader()).toTypedArray()
-                }
-                "application/json" -> {
-                    val data = Gson().fromJson(conn.inputStream.reader(UTF_8), JsonObject::class.java).getAsJsonObject("data")
-                    readCerts(data.getAsJsonPrimitive("cert").asString).toTypedArray()
-                }
-                else -> error("Invalid content-type: $ct")
-            }
-            return KeyStore.PrivateKeyEntry(kp.private, certs)
-        }
-    }
+        val resp = http.build().send(req.build(), HttpResponse.BodyHandlers.ofString())
 
-    fun getSSLContext(clientCert: KeyStore.Entry?): SSLContext {
-        val ks = KeyStore.getInstance(KeyStore.getDefaultType()).apply { load(null) }
-        clientCert?.let {
-            ks.setEntry("client-cert", clientCert, null)
-        }
-        val kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm()).apply {
-            init(ks, charArrayOf())
+        val body = resp.body()
+        if (resp.statusCode() != 200) {
+            val err = api.objectMapper.readValue(body, ApiErrorEnvelope::class.java)
+            throw ApiException(resp.statusCode(), err.error.message)
         }
 
-        var tms: Array<TrustManager>? = null
+        val controllers = listControllers(api.baseUri, pk, certs, caCerts)
 
-        if (!caCerts.isEmpty()) {
-            caCerts.forEach {
-                ks.setCertificateEntry("ca-${it.serialNumber}", it)
-            }
+        return IdentityConfig(
+            controller = api.baseUri,
+            controllers = controllers,
+            id = IdentityConfig.Id(
+                key = key,
+                cert = cert,
+                ca = caCerts.joinToString(separator = "\n"){ it.toPEM() }),)
+    }
 
-            val tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm()).apply {
-                init(ks)
+    private fun listControllers(ctrl: String, key: PrivateKey, certs: Collection<X509Certificate>, ca: Collection<X509Certificate>): List<String> {
+        val ssl = makeSSLContext(key, certs, ca)
+        val cb = ApiClient.createDefaultHttpClientBuilder().sslContext(ssl)
+        val api = ApiClient(cb, ApiClient.createDefaultObjectMapper(), ctrl)
+        val resp  = AuthenticationApi(api).authenticate("cert", null).thenApply {
+            api.requestInterceptor = Consumer { req ->
+                req.header("zt-session", it.data.token)
             }
+        }.thenCompose {
+            ControllersApi(api).listControllers(10, 0, null)
+        }.get(3, TimeUnit.SECONDS)
```

Comment:
will this log anything if there's a timeout?

Reply:
Exception is thrown